internal/receiver/awscontainerinsightreceiver/host/machineinfo.go

```diff
@@ -0,0 +1,114 @@
+package host
+
+import (
+	"sync"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/ec2metadata"
+	"github.com/aws/aws-sdk-go/aws/session"
+	cinfo "github.com/google/cadvisor/info/v1"
+	"github.com/google/cadvisor/manager"
+
+	"go.uber.org/zap"
+)
+
+type MachineInfo struct {
+	sync.RWMutex
+	logger          *zap.Logger
+	manager         manager.Manager
+	info            *cinfo.MachineInfo
+	refreshInterval time.Duration
+	instanceId      string
+	instanceType    string
+	shutdownC       chan bool
+}
+
+func NewMachineInfo(manager manager.Manager, refreshInterval time.Duration, logger *zap.Logger) *MachineInfo {
+	mInfo := &MachineInfo{
+		manager:         manager,
+		refreshInterval: refreshInterval,
+		shutdownC:       make(chan bool),
+		logger:          logger,
+	}
+
+	mInfo.refresh()
+
+	go func() {
+		refreshTicker := time.NewTicker(mInfo.refreshInterval)
+		defer refreshTicker.Stop()
+		for {
+			select {
+			case <-refreshTicker.C:
+				mInfo.refresh()
+			case <-mInfo.shutdownC:
+				return
+			}
+		}
+	}()
+
+	return mInfo
+}
+
+func (m *MachineInfo) refresh() {
+	if info, err := m.manager.GetMachineInfo(); err != nil {
+		m.logger.Error("Failed to get machine info", zap.Error(err))
+	} else {
+		m.Lock()
+		defer m.Unlock()
+		m.info = info
+	}
+
+	if m.instanceId == "" {
+		m.instanceId = m.getAwsMetadata("instance-id")
+	}
+
+	if m.instanceType == "" {
+		m.instanceType = m.getAwsMetadata("instance-type")
+	}
+}
```

Comment:
it looks these two can be merged to one call

Reply:
Right! This will reduce the api calls.